File: impls/cpp/step1_read_print.cc
#include "linenoise.hpp"
#include <iostream>
#include <memory>
#include <string>

#include "types.cc"
#include "reader.cc"
#include "printer.cc"

std::unique_ptr<Mal> READ(std::string input) {
  return read_str(input);
}

std::unique_ptr<Mal> EVAL(std::unique_ptr<Mal> input) { return std::move(input); };

std::string PRINT(std::unique_ptr<Mal> input) {
  return pr_str(std::move(input));
}

std::string rep(std::string input) {
  return PRINT(EVAL(READ(input)));
}

int main(int argc, char *argv[])
{
  linenoise::SetHistoryMaxLen(10);
  while(1) {
    std::string line;
    auto quit = linenoise::Readline("user> ", line);
    if (quit) {
      break;
    }
    Reader rdr(line);
    rdr.testprint();
    std::cout << rep(line) << std::endl;
    linenoise::AddHistory(line.c_str());
  }
  return 0;
}

File: impls/python/step1_read_print.py
#!/usr/bin/env python3
import readline
import reader as rdr
import printer as pr


def READ(input):
    return rdr.read_str(input)


def EVAL(input):
    return input


def PRINT(input):
    return pr.pr_str(input)


def rep(input):
    return PRINT(EVAL(READ(input)))


def main():
    while(True):
        try:
            line = input("user> ")
            print(rep(line))
        except EOFError:
            exit()


if __name__ == "__main__":
    main()
